### PR TITLE
feat: Implement `transformItems` API

### LINKED ISSUE #2949: Implement `transformItems` API on widgets
On every other InstantSearch flavours widgets have a simple API to perform any operation on the items before they are renderer. (ex: sort, map, ...)

From React-InstantSearch documentation:

```
Function to modify the items being displayed, e.g. for filtering or sorting them.
Takes an items as parameter and expects it back in return.

(items) => items.map(...)
```

### PR DESCRIPTION
**Pull request by francoischalifour**
This adds support for the `transformItems` option on widgets.

## How

The implementation was made at the connector level for the following reasons:

- The transformation takes place before the rendering
- Leverage the `transformItems` API in all flavors (since the connectors are/will be shared across InstantSearch libraries)

## How to review

Each commit is a widget implementation. Click on the commit and add your review there.

[See commits →](https://github.com/algolia/instantsearch.js/pull/3042/commits)

## Supported widgets

- [x] [BreadCrumb](https://community.algolia.com/instantsearch.js/v2/widgets/breadcrumb.html)
- [x] [CurrentRefinedValues](https://community.algolia.com/instantsearch.js/v2/widgets/currentRefinedValues.html)
- [x] [GeoSearch](https://community.algolia.com/instantsearch.js/v2/widgets/geoSearch.html)
- [x] [HierarchicalMenu](https://community.algolia.com/instantsearch.js/v2/widgets/hierarchicalMenu.html)
- [x] [Hits](https://community.algolia.com/instantsearch.js/v2/widgets/hits.html)
- [x] [HitsPerPageSelector](https://community.algolia.com/instantsearch.js/v2/widgets/hitsPerPageSelector.html)
- [x] [InfiniteHits](https://community.algolia.com/instantsearch.js/v2/widgets/infiniteHits.html)
- [x] [Menu](https://community.algolia.com/instantsearch.js/v2/widgets/menu.html)
- [x] [MenuSelect](https://community.algolia.com/instantsearch.js/v2/widgets/menuSelect.html)
- [x] [NumericRefinementList](https://community.algolia.com/instantsearch.js/v2/widgets/numericRefinementList.html)
- [x] [NumericSelector](https://community.algolia.com/instantsearch.js/v2/widgets/numericSelector.html)
- [x] [RefinementList](https://community.algolia.com/instantsearch.js/v2/widgets/refinementList.html)
- [x] [SortBySelector](https://community.algolia.com/instantsearch.js/v2/widgets/sortBySelector.html)

These widgets have been selected based on whether they contain items and based on the current React InstantSearch API. Let me know if this list should be updated.

[See all InstantSearch.js widgets →](https://community.algolia.com/instantsearch.js/v2/widgets.html)

<details>
  <summary>React InstantSearch widgets implementing <code>transformItems</code></summary>

- Breadcrumb
- ClearRefinements
- CurrentRefinements
- HierarchicalMenu
- HitsPerPage
- Menu
- MenuSelect
- NumericMenu
- RefinementList
- SortBy

</details>

---

Closes #2949

### CODE REVIEW

**Comment by francoischalifour** [comment]
Is this list correct?

React InstantSearch implements the `transformItems` API for these widgets although I'm not convinced they're useful in InstantSearch.js:

* ClearRefinedValues
* CurrentRefinements
* HitsPerPageSelector
* NumericRefinementList
* NumericSelector
* RefinementList
* SortBy

**Comment by francoischalifour** [comment]
Ready for review @algolia/instantsearch-for-websites.

I can cherry pick my commits to create several PRs if it makes it easier.

File: src/connectors/hits-per-page/connectHitsPerPage.js [hunk 1]
```diff
@@ -192,7 +194,7 @@ The first one will be picked, you should probably set only one default value`
         );
       },
 
-      _transformItems({ hitsPerPage }) {
+      _normalizeItems({ hitsPerPage }) {
```
Comment (bobylito) [inline src/connectors/hits-per-page/connectHitsPerPage.js]: Yes good move 👍

File: src/connectors/hits/connectHits.js [hunk 2]
```diff
@@ -77,8 +79,14 @@ export default function connectHits(renderFn, unmountFn) {
     },
 
     render({ results, instantSearchInstance }) {
-      if (widgetParams.escapeHits && results.hits && results.hits.length > 0) {
-        results.hits = escapeHits(results.hits);
+      if (results.hits && results.hits.length > 0) {
+        if (typeof widgetParams.transformItems === 'function') {
```
Comment (bobylito) [inline src/connectors/hits/connectHits.js]: Why not provide a default implementation (identity) like in other widget and remove this if statement? Also I'd suggest to test the type sooner at the initialization.

File: src/connectors/infinite-hits/connectInfiniteHits.js [hunk 3]
```diff
@@ -110,12 +112,14 @@ export default function connectInfiniteHits(renderFn, unmountFn) {
...
         }
 
-        if (
-          widgetParams.escapeHits &&
-          results.hits &&
-          results.hits.length > 0
-        ) {
-          results.hits = escapeHits(results.hits);
+        if (results.hits && results.hits.length > 0) {
+          if (typeof widgetParams.transformItems === 'function') {
```
Comment (bobylito) [inline src/connectors/infinite-hits/connectInfiniteHits.js]: Same remarks than in the hits connector.

File: src/connectors/menu/__tests__/connectMenu-test.js [hunk 4]
```diff
@@ -230,6 +230,59 @@ describe('connectMenu', () => {
...
+      attributeName: 'category',
+      transformItems: items =>
+        items.map(item => ({
+          ...item,
+          label: 'transformed',
+        })),
+    });
+
+    const helper = jsHelper({}, '', widget.getConfiguration({}));
+    helper.search = sinon.stub();
```
Comment (bobylito) [inline src/connectors/menu/__tests__/connectMenu-test.js]: When we touch a test file that is still using sinon, we try to take the opportunity to replace it with Jest mocks. Can you do that? 🙏

Comment (francoischalifour) [reply src/connectors/menu/__tests__/connectMenu-test.js]: For sure! I initially didn't want to edit too much code as it was intended to be merged on the yet-to-be consequent `feat/3.0` branch.

File: src/connectors/hits/connectHits.js (hunk 2 above)
Comment (francoischalifour) [reply src/connectors/hits/connectHits.js]: Glad you caught that. I wasn't sure if I should have updated the code flow of the connector or should have adapted my code on the existing flow.

**Comment by francoischalifour** [comment]
Tests have been updated but rely on #3045.

**Comment by bobylito** [comment]
> Tests have been migrated to Jest but rely on #3045.

merged :)

**Pull request merged**